File: 2025/07/solution_2.py
from sys import argv

with open(argv[1]) as f:
    lines = list(map(list, f.read().strip().replace("S", "|").split("\n")))

timeline_counts = [1 if val == "|" else 0 for val in lines[0]]
for r, line in enumerate(lines[1:], start=1):
    new_timeline_counts = [0] * len(timeline_counts)
    for c, char in enumerate(line):
        if lines[r - 1][c] == "|":
            if char == "^":
                line[c - 1] = "|"
                new_timeline_counts[c - 1] += timeline_counts[c]
                line[c + 1] = "|"
                new_timeline_counts[c + 1] += timeline_counts[c]
            else:
                line[c] = "|"
                new_timeline_counts[c] += timeline_counts[c]
    timeline_counts = new_timeline_counts

print(sum(timeline_counts))

File: 2025/07/src/solution_2.rs
use crate::common::GridVal;

pub fn solve(grid: &Vec<Vec<GridVal>>) {
    let mut prev_row: Vec<GridVal> = grid.first().expect("Input has no lines!").clone();
    let mut timeline_counts = vec![0u64; prev_row.len()];
    prev_row.iter()
        .enumerate()
        .find_map(|(i, val)| if *val == GridVal::Beam { Some(i) } else { None })
        .map(|i| timeline_counts[i] = 1);

    for new_row in grid.iter().skip(1) {
        let mut new_row: Vec<GridVal> = new_row.clone();
        let mut new_timeline_counts = vec![0u64; prev_row.len()];

        for i in prev_row.iter().enumerate().filter_map(|(i, val)| {
            if *val == GridVal::Beam {
                Some(i)
            } else { None }
        }) {
            if new_row[i] == GridVal::Splitter {
                assert!(i > 0, "Attempted to split the beam off the left edge of the matrix");
                assert!(i < new_row.len(), "Attempted to split the beam off the left edge of the matrix");

                new_row[i-1] = GridVal::Beam;
                new_timeline_counts[i-1] += timeline_counts[i];
                new_row[i+1] = GridVal::Beam;
                new_timeline_counts[i+1] += timeline_counts[i];
            } else {
                new_row[i] = GridVal::Beam;
                new_timeline_counts[i] += timeline_counts[i];
            }
        }

        prev_row = new_row;
        timeline_counts = new_timeline_counts;
    }

    let total_timelines: u64 = timeline_counts.iter().sum();
    println!("There will be {total_timelines} timelines!")
}
